pkgs/fc/agent/fc/util/channel.py

```diff
@@ -116,7 +119,38 @@ def switch(self, lazy=True, show_trace=False):
         nixos.register_system_profile(self.system_path, self.log)
         # New system is registered, delete the temporary result link.
         os.unlink(out_link)
-        return nixos.switch_to_system(self.system_path, lazy, self.log)
+
+        current_release = nixos.get_release_version(
+            nixos.running_system_version()
+        )
+        next_release = nixos.get_release_version(
+            (Path(self.system_path) / "nixos-version").read_text()
+        )
+
+        if current_release != next_release:
+            reboot_delay = 10
+            self.log.warn(
+                "release-change-requires-reboot",
+                current_release=current_release,
+                next_release=next_release,
+            )
+            while reboot_delay:
+                self.log.warn(
+                    "reboot-scheduled",
+                    _replace_msg=f"WILL REBOOT IN {reboot_delay} SECONDS. PRESS Ctrl-C TO ABORT.",
+                )
+                time.sleep(1)
+                reboot_delay -= 1
+            nixos.switch_to_system(self.system_path, lazy, "boot", self.log)
```

Comment:
This should be
```
if nixos.switch_to_system(switch_path, lazy, "boot", self.log) is False:
  return False
```
to prevent reboot when a reboot when no config change happened. This is a case that should not happen tho.